src/drivers/dma/bcm283x/bcm283x_dma.c

```diff
@@ -0,0 +1,222 @@
+    mem->bus_addr = vc_mem_lock(mem->mb_handle);
+    mem->physical_addr = DMA_BUS_TO_PHYS(mem->bus_addr);
+    mem->size = size;
+
+    assert(mem->bus_addr != 0);
+    return mem;
+}
+
+// Unlock and deallocate memory
+//
+void dma_free(Dma_mem_handle *mem)
+{
+    if (mem->physical_addr == NULL)
+        return;
+
+    vc_mem_unlock(mem->mb_handle);
+    vc_mem_free(mem->mb_handle);
+    mem->physical_addr = NULL;
+    free(mem);
+}
+
+// Enable and reset DMA
+int bcm283x_dma_config(int dma_chan) {
+    assert(dma_chan <= DMA_CHANNELS);
+
+    *(REGS_DMA_ENABLE) |= ( 1 << dma_chan );
+    REGS_DMA(dma_chan)->cs = DMA_CS_RESET;
+    return 0;
+}
+
+// Start DMA - basic memory copy
+// dma_chan - integer 0-15
+// dst, src - are destination and source memory locations in bus address space
+// words - 4-byte aligned memory length
+int bcm283x_dma_transfer(int dma_chan, uint32_t dst, uint32_t src, int words) {   
+    assert(dma_chan <= DMA_CHANNELS);
+
+    // A standard copy block to support basic implementation
+    shared_conbk.source_ad = src;
+    shared_conbk.dest_ad = dst;
+    shared_conbk.txfr_len = words * sizeof(uint32_t);
+    shared_conbk.ti = DMA_TI_SRC_INC | DMA_TI_DEST_INC;
+    shared_conbk.nextconbk = 0;
+    shared_conbk.stride = 0;
+    shared_conbk.debug = 0;
+
+    return bcm283x_dma_transfer_conbk(dma_chan, &shared_conbk);
+}
+
+// Start DMA
+// dma_chan - integer 0-15
+// conbk - pointer to data type in physical memory
+int bcm283x_dma_transfer_conbk(int dma_chan, volatile Dma_conbk *conbk) {
+    assert(dma_chan <= DMA_CHANNELS);
+
+    REGS_DMA(dma_chan)->conblk_ad = (uint32_t)conbk;  
+    REGS_DMA(dma_chan)->cs = DMA_CS_END;    // Clear 'end' flag
+    REGS_DMA(dma_chan)->debug = DMA_DEBUG_READ_ERROR 
+        | DMA_DEBUG_FIFO_ERROR 
+        | DMA_DEBUG_READ_LAST_NOT_SET_ERROR; // clear error bits
+    REGS_DMA(dma_chan)->cs = DMA_CS_ACTIVE; // Start DMA
+    return 0;
+}
+
+int bcm283x_dma_in_progress(int dma_chan) {
+    assert(dma_chan <= DMA_CHANNELS);
+
+    return (REGS_DMA(dma_chan)->cs & DMA_CS_ACTIVE);
+}
+
+// Display DMA registers (for debugging)
+void bcm283x_dma_disp(int dma_chan)
+{
+    assert(dma_chan <= DMA_CHANNELS);
+
+    printk("\ncs(%d) %08X\n", dma_chan, REGS_DMA(dma_chan)->cs);
+    printk("conblk_ad(%d) %08X\n", dma_chan, REGS_DMA(dma_chan)->conblk_ad);
+    printk("ti(%d) %08X\n", dma_chan, REGS_DMA(dma_chan)->ti);
+    printk("source_ad(%d) %08X\n", dma_chan, REGS_DMA(dma_chan)->source_ad);
+    printk("dest_ad(%d) %08X\n", dma_chan, REGS_DMA(dma_chan)->dest_ad);
+    printk("txfr_len(%d) %08X\n", dma_chan, REGS_DMA(dma_chan)->txfr_len);
+    printk("stride(%d) %08X\n", dma_chan, REGS_DMA(dma_chan)->stride);
+    printk("nextconbk(%d) %08X\n", dma_chan, REGS_DMA(dma_chan)->nextconbk);
+    printk("debug(%d) %08X\n", dma_chan, REGS_DMA(dma_chan)->debug);
+}
+
+static int bcm283x_dma_init(void) {
+	int ret = 0;
+    
+#if 0
```

Comment:
Interrupts do not use now?

Reply:
Yes, still have to figure that out.  and you'd only want to configure an interrupt on specific settup with a DMA channel for the driver that is using it so, but that leads to other quiestion about DMA, is it warrent a general embox driver model.

Reply:
Will remove for now and add back in either SPI driver use and/or in more general DMA device.  Probably specific example with SPI usage first and then generalze from there.